### DISCUSSION TITLE
Load `router.query` synchronously upon page load

### DISCUSSION THREAD

**Discussion by heymartinadams** [vercel/next.js#18353] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Currently, NextJS’ `const router = useRouter()` only produces a URL’s query via the `router.query` a second or so after load, necessitating a state update before the information can be accessed by the page.

For example, a link such as `/event?option=rsb2&qty=2` won’t give me the `option` and `qty` query right away, even though that information is already included in `router.asPath`.

## Describe the solution you'd like

I’d like to have instant access to the `router.query` parameter upon page load.

## Describe alternatives you've considered

Because we require that information instantly upon load, we instead have to do a workaround whereby we parse the query ourselves from `router.asPath` each time:

```js
// Returns array of search keys and values. For example, this `/events?level=introductory&topic=relationships,communication`... creates an array of key/value pairs, `[{ level: 'introductory' }, { topic: 'relationships,communication' }]`.

const getUrlParams = str =>
  str && str.includes('?')
    ? str
        .split('?')[1]
        .split('&')
        .map(item => ({
          [item.split('=')[0]]: [
            // For `single=true`, make value boolean `true`
            item.split('=')[1] === 'true' ? true : null,
            // For `single=false`, make value boolean `false`
            item.split('=')[1] === 'false' ? false : null,
            // For `level=introductory`, return string value, otherwise, if `single` (with no `=value`) return `single` as `true`
            item.split('=')[1] ? item.split('=')[1] : true
          ]
            // And find the first possibility that’s either `true`, `false`, or a string value — in other words, that’s not `null`
            .find(x => x === true || x === false || x)
        }))
    : []
```

## Additional context

Here’s a screen video that logs both `router.query` and our own solution. As you can see, our solution captures the query synchronously upon page load and produces an array of length 2 even as `router.query` still only shows the `eventId`.

![Screen Recording 2020-10-27 at 4 41 14 PM](https://user-images.githubusercontent.com/11673745/97374072-ae088900-1874-11eb-91a5-88f76e0cd47c.gif)

**Comment by megetron** [comment]
you can use this instead
```
const params = Object.fromEntries(new URLSearchParams(location.search));
```

**Reply by zeeshanibnali** [reply]
This saved me a lot of time brother. Thanks a lot!
